### PR TITLE
[Feat.] Add AS Lifecycle Hook

### PR DESCRIPTION
**Pull request by muneeb-jan**
### What this PR does / why we need it
Adds [AS lifecycle hooks](https://docs.otc.t-systems.com/auto-scaling/api-ref/apis/lifecycle_hooks/index.html) to GopherTelekomCloud.

### Which issue this PR fixes

Fixes issue #731 

### Tests Performed

```
=== RUN   TestLifecycleHooksLifecycle
    helpers.go:73: Attempting to create AutoScaling Configuration
    helpers.go:98: Created AutoScaling Configuration: c582c684-bd51-484b-ae57-42f24945fc8b
    helpers.go:48: Attempting to create AutoScaling Group
    helpers.go:51: Created AutoScaling Group: 99c09047-3022-4744-86bb-7c35adf3e85d
    lifecycle_hooks_test.go:27: Attempting to create Topic: as-lifecycle-hooks-topic-VLV
    lifecycle_hooks_test.go:39: Attempting to create Lifecycle Hook
    lifecycle_hooks_test.go:42: Created Lifecycle Hook: as-lifecycle-hook-create-86u
    lifecycle_hooks_test.go:52: Attempting to update Lifecycle Hook
    lifecycle_hooks_test.go:57: Updated Lifecycle Hook: as-lifecycle-hook-create-86u
    lifecycle_hooks_test.go:59: Listing all Lifecycle Hooks
    tools.go:74: {
          "lifecycle_hook_name": "as-lifecycle-hook-create-86u",
          "lifecycle_hook_type": "INSTANCE_TERMINATING",
          "default_result": "ABANDON",
          "default_timeout": 4800,
          "notification_topic_urn": "urn:smn:eu-de:7b697e01b13e431aa3f4c8a7afe2046f:as-lifecycle-hooks-topic-VLV",
          "notification_topic_name": "as-lifecycle-hooks-topic-VLV",
          "notification_metadata": "",
          "create_time": "2024-10-07T13:10:06Z"
        }
    lifecycle_hooks_test.go:67: Attempting to delete Lifecycle Hook
    lifecycle_hooks_test.go:70: Deleted Lifecycle Hook: as-lifecycle-hook-create-86u
    helpers.go:122: Attempting to Delete Topic: urn:smn:eu-de:7b697e01b13e431aa3f4c8a7afe2046f:as-lifecycle-hooks-topic-VLV
    helpers.go:127: Deleted Topic: urn:smn:eu-de:7b697e01b13e431aa3f4c8a7afe2046f:as-lifecycle-hooks-topic-VLV
    helpers.go:60: Attempting to delete AutoScaling Group
    helpers.go:65: Deleted AutoScaling Group: 99c09047-3022-4744-86bb-7c35adf3e85d
    helpers.go:103: Attempting to delete AutoScaling Configuration
    helpers.go:106: Deleted AutoScaling Configuration: c582c684-bd51-484b-ae57-42f24945fc8b
--- PASS: TestLifecycleHooksLifecycle (12.01s)
PASS
```

### CODE REVIEW

**Review by anton-sidelnikov — CHANGES_REQUESTED** [review]
Tests?

**Comment by muneeb-jan** [comment]
> Tests?

This one is not up for review yet.

File: acceptance/openstack/autoscaling/v1/lifecycle_hooks_test.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		autoscaling.DeleteAutoScalingGroup(t, client, groupID)
+	})
+
+	topicName := tools.RandomString("as-lifecycle-hooks-topic-", 3)
+	t.Logf("Attempting to create Topic: %s", topicName)
+	topicURN, err := autoscaling.GetNotificationTopicURN(topicName)
+	if err != nil {
+		t.Logf("Error while creating the notification topic: %s", topicName)
+	}
+	defer autoscaling.DeleteTopic(t, topicURN)
```
Comment (anton-sidelnikov) [inline acceptance/openstack/autoscaling/v1/lifecycle_hooks_test.go]: why not cleanup?

File: acceptance/openstack/autoscaling/v1/lifecycle_hooks_test.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	lifecycleHooks, err := lifecyclehooks.List(client, groupID)
+	th.AssertNoErr(t, err)
+	for _, lifecycleHook := range lifecycleHooks {
+		tools.PrintResource(t, lifecycleHook)
+	}
+
+	t.Logf("Attempting to delete Lifecycle Hook")
+	err = lifecyclehooks.Delete(client, groupID, lifecycleHookName)
+	th.AssertNoErr(t, err)
+	t.Logf("Deleted Lifecycle Hook: %s", lifecycleHookName)
```
Comment (anton-sidelnikov) [inline acceptance/openstack/autoscaling/v1/lifecycle_hooks_test.go]: better to put this also in Cleanup

File: openstack/autoscaling/v1/lifecycle_hooks/Create.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	// ABANDON:
+	// If an instance is starting, ABANDON indicates that your customized operations failed, and the instance will be terminated.
+	// In such a case, the scaling action fails, and you must create a new instance.
+	// If an instance is stopping, ABANDON allows instance termination BUT stops other lifecycle hooks.
+	// CONTINUE:
+	// If an instance is starting, CONTINUE indicates that your customized operations are successful and the instance can be used.
+	// If an instance is stopping, CONTINUE allows instance termination AND the completion of other lifecycle hooks.
+	DefaultResult string `json:"default_result,omitempty"`
+	// Specifies the lifecycle hook timeout duration, which ranges from 60 to 86400 seconds. The default value is 3600.
+	DefaultTimeout int `json:"default_timeout,omitempty"`
```
Comment (anton-sidelnikov) [inline openstack/autoscaling/v1/lifecycle_hooks/Create.go:28]: use here pointer, without pointer impossible to omit this parameter, could lead to error

File: openstack/autoscaling/v1/lifecycle_hooks/Update.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	// ABANDON:
+	// If an instance is starting, ABANDON indicates that your customized operations failed, and the instance will be terminated.
+	// In such a case, the scaling action fails, and you must create a new instance.
+	// If an instance is stopping, ABANDON allows instance termination BUT stops other lifecycle hooks.
+	// CONTINUE:
+	// If an instance is starting, CONTINUE indicates that your customized operations are successful and the instance can be used.
+	// If an instance is stopping, CONTINUE allows instance termination AND the completion of other lifecycle hooks.
+	DefaultResult string `json:"default_result,omitempty"`
+	// Specifies the lifecycle hook timeout duration, which ranges from 60 to 86400 seconds. The default value is 3600.
+	DefaultTimeout int `json:"default_timeout,omitempty"`
```
Comment (anton-sidelnikov) [inline openstack/autoscaling/v1/lifecycle_hooks/Update.go:26]: and here pointer

**Review by anton-sidelnikov — CHANGES_REQUESTED** [review]

File: acceptance/openstack/autoscaling/v1/lifecycle_hooks_test.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/opentelekomcloud/gophertelekomcloud/acceptance/tools"
+	lifecyclehooks "github.com/opentelekomcloud/gophertelekomcloud/openstack/autoscaling/v1/lifecycle_hooks"
+	th "github.com/opentelekomcloud/gophertelekomcloud/testhelper"
+)
+
+func TestLifecycleHooksLifecycle(t *testing.T) {
+	client, err := clients.NewAutoscalingV1Client()
+	th.AssertNoErr(t, err)
+
+	networkID := clients.EnvOS.GetEnv("NETWORK_ID")
```
Comment (artem-lifshits) [inline acceptance/openstack/autoscaling/v1/lifecycle_hooks_test.go:15]: Please put this check before client init.
If env vars are missing the client init becomes redundant.

File: acceptance/openstack/autoscaling/v1/lifecycle_hooks_test.go (hunk 2 above)
Comment (muneeb-jan) [reply acceptance/openstack/autoscaling/v1/lifecycle_hooks_test.go]: done

File: acceptance/openstack/autoscaling/v1/lifecycle_hooks_test.go (hunk 1 above)
Comment (muneeb-jan) [reply acceptance/openstack/autoscaling/v1/lifecycle_hooks_test.go]: done

File: openstack/autoscaling/v1/lifecycle_hooks/Create.go (hunk 3 above)
Comment (muneeb-jan) [reply openstack/autoscaling/v1/lifecycle_hooks/Create.go:28]: default values in api

**Review by anton-sidelnikov — APPROVED** [review]

**Review by artem-lifshits — APPROVED** [review]

**Pull request merged**